File: hw3/tree.py
#Jiazhou Liu ljzhou@bu.edu

class Tree:
    def __init__(self):
        self.value=None
        self.left=None
        self.right=None
        
    def __eq__(self,other):
        if (self.value==None)&(other.value==None):
            return True
        elif(self.value==None)&(other.value!=None):
            return False
        elif(self.value!=None)&(other.value==None):
            return False
        else:
            return (self.value==other.value)&(self.left==other.left)&(self.right==other.right)
    
    def leafcount(self):
        if (self.value==None):
            return 1
        else:
            return self.left.leafcount()+self.right.leafcount()
            
    def nodecount(self):
        if (self.value==None):
            return 0
        else:
            return self.left.nodecount()+self.right.nodecount()+1
    def height(self):
        if (self.value==None):
            return 0
            
        else:
            return max(self.left.height()+1,self.right.height()+1)
    def perfect(self):
        if (self.value==None):
            return True
        elif(self.left.height()!=self.right.height()):
             return False
        else:
            return self.left.perfect()and self.right.perfect()

    def degenerate(self):
        if (self.value==None):
            return True
        elif(self.left.value!=None)and(self.right.value!=None):
            return False
        else:
            return self.left.degenerate()and self.right.degenerate()

    def list(self):
        if (self.value==None):
            return []
        elif(self.degenerate()==False):
            return []
        else:
            return [self.value]+self.left.list()+self.right.list()
            
        
def Node(value,left,right):
    n=Tree()
    n.value=value
    n.left=left
    n.right=right
    return n
def Leaf():
    n=Tree()
    return n    
    

File: hw3/Tree.hs
--Jiazhou Liu ljzhou@bu.edu
module Tree where

data Tree a = Leaf | Node a (Tree a) (Tree a) deriving (Eq, Show)


	
mapT f Leaf = Leaf
mapT f (Node a l r)= Node (f a) (mapT f l) (mapT f r)

foldT f b Leaf = b
foldT f b (Node a l r)= f a (foldT f b l) (foldT f b r)

add3  m n p =n+p
leafCount Leaf= 1
leafCount (Node a l r)= foldT add3 1 (Node a l r)

add2 m n p = n+p+1

nodeCount Leaf= 0
nodeCount (Node a l r)=foldT add2 0 (Node a l r)

add1 m n p = n + 1
add0 m n p = p + 1

height Leaf=0
height (Node a l r)=max (foldT add1 0 (Node a l r)) (foldT add0 0 (Node a l r))

perfect Leaf= True
perfect (Node a l r)
	|(height l /= height r)=False
	|otherwise = (perfect l)&&(perfect r)
	
degenerate Leaf=True
degenerate (Node a l r)
	|(l/=Leaf)&&(r/=Leaf) = False
	|otherwise = (degenerate l)&&(degenerate r)

add5 a b c = [a]++b++c
	
list Leaf = []
list (Node a l r)
	|(degenerate (Node a l r)==False) = []
	|otherwise = foldT add5 [] (Node a l r)
